src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java

```diff
@@ -184,4 +194,20 @@ public void invalidateUserJwt(final Cookie customJWTCookie, final HttpServletRes
         thirdPartyAuthToken.setMaxAge(0);
         response.addCookie(thirdPartyAuthToken);
     }
+
+    public void validateRoles(List<Role> userRoles, String payloadRoles) throws UnauthorizedException {
+        final List<String> formattedPayloadRoles = removeSquareBracketsAndTrim(Arrays.asList(payloadRoles.split(",")));
+        final Set<String> formattedUserRoles = userRoles.stream()
+                .map(role -> roleMapper.roleToRoleDto(role).getName())
+                .collect(Collectors.toSet());
+
+        if(!formattedUserRoles.containsAll(formattedPayloadRoles)){
```

Comment:
This check doesn't cover all scenarios. Just ensures formattedPayloadRoles is a subset of formattedUserRoles. What about the inverse scenario?

Reply:
Ideally add unit tests that would capture this scenario too. Along with any other edge cases that arent already tested

Reply:
If both collections are a set you can check if they are equal to each other instead of containsAll.  As Dom said we need unit tests to cover all scenarios

Reply:
added 4 tests for this which should cover everything